api/lock/lock.go

```diff
@@ -0,0 +1,6 @@
+package lock
+
+type Lock interface {
```

Comment:
[just my opinion] the design of the `Lock` interface seems very straightforward to me, and it's in line with typical Go interface design in that a `Lock` is simply anything that can be acquired or released. By that same token, defining things like expiry time in the implementation gives you a lot of flexibility and doesn't clutter up the general idea of a `Lock`.

I've found this article helpful in the past, and it greatly informed how I looked at the usage of interfaces in Go: This is a core concept in Go’s type system; instead of designing our abstractions in terms of what kind of data our types can hold, we design our abstractions in terms of what actions our types can execute.

Key idea here (for me) is:
"instead of designing our abstractions in terms of what kind of data our types can hold, we design our abstractions in terms of what actions our types can execute."

Reply:
Sure I agree. I used interfaces to describe the general pattern here though not necessarily advocating for that interface. I.e. 
```go
//instead of:
lock := lock.NewLock(...)
lock.Acquire("1")
lock.Release("1")

//have:
lock := DistLockService.NewLock("1", ...)
lock.Acquire()
lock.Release() //will only release the lock for id "1" if it acquired in the first place
```

Subtle change but a scenario where it might be helpful would be where you might release an expired lock after it was acquired again.
```go
lock := lock.NewLock(...) //expires after 5 minutes
lock.Acquire("1")
// do somekind of long running operation till 5 minutes expires. lock will be released.

// a new operation begins. lock acquired on the same id again in a different operation
lock.Acquire("1")

// a bug in the code calls this again because nothing prevents that from happening after 6 minutes from the first operation
lock.Release("1")

// not the previous release called after 6 minutes would release the newly acquired lock after 1 minute.
```

So my comment was more for having immutable lock objects that can only be used once. But it is just a nit though.
```

Reply:
```
// a new operation begins. lock acquired on the same id again in a different operation
lock.Acquire("1")
```
Wouldn't that `Acquire` fail because the initial one hasn't released it yet?

Reply:
My first implementation actually was that design:
```
lock := lock.NewLock(...)
lock.Acquire()
lock.Release()
```

Which is great when dealing with locks for each Daemon/Janitor since we can just create them in main. However, this doesn't work well when we bring jobs into the equation since each job's lock needs to created on the fly. We'd need to introduce a `LockFactory`, etc. etc.

Reply:
> In the scenario you are describing, I'm not sure how the other design would avoid that?

The lock schema would look like this
```go
type LockSchema struct {
  LockID         string
  Acquired     int64
  AcquiredBy string //uuid
}
```

so each time a lock.Acquire is called, a new uuid is generated. This would allow for a lock to release only if it acquired it.

Reply:
what if it expired though?

Reply:
It would still be released the same as it does now on expiration. The uuid is only for ensuring the instance of lock that acquired the lock can release it. This is just a [design nit] though.